src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,353 @@
+  (disj : disjoint A B) :
+  ∃ ε : ℝ, 0 < ε ∧ disjoint (A + metric.ball 0 ε) (B + metric.ball 0 ε) :=
+begin
+  obtain ⟨u, -, u_pos, u_lim⟩ := exists_seq_strict_anti_tendsto (0 : ℝ),
+  suffices h : ∃ (n : ℕ), disjoint (A + metric.ball 0 (u n)) (B + metric.ball 0 (u n)),
+  { obtain ⟨n, hn⟩ := h,
+    exact ⟨u n, u_pos n, hn⟩ },
+  by_contra' h,
+  simp only [not_disjoint_iff, set.mem_add, metric.mem_ball, dist_zero_right,
+    ← exists_and_distrib_left, ← exists_and_distrib_right, and_assoc] at h,
+  choose z f f' g g' h₁ h₂ h₃ h₄ h₅ h₆ using h,
+  obtain ⟨w, hw, φ, hφ₁, hφ₂ : tendsto (f ∘ _) _ _⟩ := hA.tendsto_subseq h₁,
+  refine disj ⟨hw, mem_of_is_closed_sequential hB (λ n, h₄ (φ n)) _⟩,
+  suffices hfg : tendsto (f - g) at_top (𝓝 0),
+  { simpa only [sub_sub_cancel, sub_zero, comp_app, pi.sub_apply]
+      using hφ₂.sub (hfg.comp hφ₁.tendsto_at_top) },
+  suffices : ∀ n, ∥(f - g) n∥ ≤ 2 * u n,
+  { apply squeeze_zero_norm this,
+    rw ←mul_zero (2 : ℝ),
+    exact u_lim.const_mul (2:ℝ) },
+  intro n,
+  have : f n - g n = g' n - f' n,
+  { rw [sub_eq_iff_eq_add', ←add_sub_assoc, h₆, ←h₃, add_sub_cancel] },
+  rw [pi.sub_apply, this, two_mul],
+  exact (norm_sub_le _ _).trans (add_le_add (h₅ n).le (h₂ n).le),
+end
+
+variables [normed_space ℝ E]
+
+/-- Given a set `C` which is a convex neighbourhood of `0` and a point `x₀` outside of it, there is
+a continuous linear functional `f` separating `x0` and `C`, in the sense that it sends `x₀` to 1 and
+all of `C` to values strictly below `1`. -/
+lemma separate_convex_open_set {C : set E} (zero_mem : (0:E) ∈ C) (hC : convex ℝ C)
+  (hC₂ : is_open C) {x₀ : E} (hx₀ : x₀ ∉ C) :
+  ∃ (f : E →L[ℝ] ℝ), f x₀ = 1 ∧ ∀ x ∈ C, f x < 1 :=
+begin
+  let f : linear_pmap ℝ E ℝ :=
+    linear_pmap.mk_span_singleton x₀ 1 (ne_of_mem_of_not_mem zero_mem hx₀).symm,
+  have hfx₀ : f ⟨(1:ℝ) • x₀, by { dsimp, rw submodule.mem_span_singleton, exact ⟨1, rfl⟩ }⟩ = 1,
+  { simp_rw [linear_pmap.mk_span_singleton_apply, one_smul] },
+  obtain ⟨φ, hφ₁, hφ₂⟩ := exists_extension_of_le_sublinear f (gauge C) _ _ _,
+  { refine ⟨⟨φ, (φ.to_add_monoid_hom.uniform_continuous_of_continuous_at_zero _).continuous⟩, _, _⟩,
+    { refine continuous_at_zero_of_exists_open _ (λ ε hε, ⟨(ε • C) ∩ (-ε • C), ⟨_, _⟩, _, _⟩),
+      { exact mem_smul_set.mpr ⟨0, zero_mem, smul_zero _⟩ },
+      { exact mem_smul_set.mpr ⟨0, zero_mem, smul_zero _⟩ },
+      { exact (is_open_map_smul₀ hε.ne' _ hC₂).inter
+          (is_open_map_smul₀ (neg_ne_zero.mpr hε.ne.symm) _ hC₂), },
+      rintro x ⟨hx₁, hx₂⟩,
+      rw [real.norm_eq_abs, abs_lt, neg_lt, ←linear_map.map_neg],
+      split; apply (hφ₂ _).trans_lt,
+      { refine gauge_lt_of_mem_smul (-x) ε hε zero_mem hC hC₂ _,
+        rw [mem_smul_set_iff_inv_smul_mem₀ hε.ne', smul_neg],
+        rwa [mem_smul_set_iff_inv_smul_mem₀ (neg_ne_zero.mpr hε.ne'), inv_neg, neg_smul] at hx₂, },
+      { exact gauge_lt_of_mem_smul x ε hε zero_mem hC hC₂ hx₁, } },
```

Comment:
How to do this without `continuous_at_zero_of_exists_open`:
```suggestion
    { change tendsto _ _ _,
      rw (nhds_basis_opens (0:E)).tendsto_iff metric.nhds_basis_ball,
      refine λ ε hε, ⟨(ε • C) ∩ (-ε • C), ⟨⟨_, _⟩, _⟩, _⟩,
      { exact mem_smul_set.mpr ⟨0, zero_mem, smul_zero _⟩ },
      { exact mem_smul_set.mpr ⟨0, zero_mem, smul_zero _⟩ },
      { exact (is_open_map_smul₀ hε.ne' _ hC₂).inter
          (is_open_map_smul₀ (neg_ne_zero.mpr hε.ne.symm) _ hC₂), },
      rintro x ⟨hx₁, hx₂⟩,
      have : ∥φ x∥ < ε,
      { rw [real.norm_eq_abs, abs_lt, neg_lt, ←linear_map.map_neg],
        split; apply (hφ₂ _).trans_lt,
        { refine gauge_lt_of_mem_smul (-x) ε hε zero_mem hC hC₂ _,
          rw [mem_smul_set_iff_inv_smul_mem₀ hε.ne', smul_neg],
          rwa [mem_smul_set_iff_inv_smul_mem₀ (neg_ne_zero.mpr hε.ne'), inv_neg, neg_smul] at hx₂ },
        { exact gauge_lt_of_mem_smul x ε hε zero_mem hC hC₂ hx₁, } },
      simp [this] },
```

Reply:
Ahh this is great to know! So the idea is to avoid picking elements of the filter manually to show `tendsto` and instead use `has_basis.tendsto_iff` and use convenient filter bases? We still keep the line
```
{ refine ⟨⟨φ, (φ.to_add_monoid_hom.uniform_continuous_of_continuous_at_zero _).continuous⟩, _, _⟩,
```
though right? Since that's the one which lets us only consider 0

Reply:
Sorry, yup :)